Integrate Static File Routes

nZac opened the pull request:
There needs to be some test written for this, however, with October growing long and my time being short I find it prudent to propose something.

The one change I wanted to make but don't have the time to, instead of using the prefixing, using a regular expression to match a path after the prefix. My guess is it could be done, but my expertise in the area of regex is limited at best.

LMK what else I am missing, likely docs?

kgriffs commented on falcon/statics.py:
```diff
@@ -0,0 +1,43 @@
+
+        resp.content_type = self.mime_types.get(suffix, 'application/octet-stream')
+
+        raise falcon.HTTPStatus('200 OK')
```
No need to raise. The framework already defaults to 200 OK (and simply setting `resp.status` is preferred in any case).

kgriffs commented on falcon/statics.py:
```diff
@@ -0,0 +1,43 @@
+
+    def __call__(self, req, resp):
+        stripped = req.path.replace(self.prefix, '')
+        cleaned = stripped.replace('../', '').lstrip('/')
```
It may be better to instead run `stripped` through `os.path.normpath()` and then explicitly raise `falcon.HTTPNotFound()` if the result starts with `'..'` or `/` (rather than checking first if the file exists with those characters stripped). This has the added benefits of allowing harmless redundant separators and level references, replacing forward slashes with backslashes on Windows, and ensuring that malformed requests aren't inadvertently mapped to existing files.

kgriffs commented on falcon/statics.py:
```diff
@@ -0,0 +1,43 @@
+    def __init__(self, prefix, parent_folder):
+        self.prefix = prefix
+        self.parent_folder = os.path.abspath(parent_folder)
+        self.mime_types = copy.copy(mimetypes.encodings_map)
```
Also, I think it would be helpful to factor this out to initialize a new option, `falcon.ResponseOptions.static_media_types`, and use that here. Then, people can add additional types to that dict as needed.

nZac commented:
@kgriffs I updated as requested and add a test. The tests in this lib have a variety of different methods so I did what felt comfortable for me, if it isn't to standard let me know.

Also, I am not sure where to plug this in now that I converted to using a list of static routes on the `API` instance. My guess would be that `_get_responder` would be the best place.

Docs are still a question mark as well.

kgriffs commented on falcon/api.py:
```diff
@@ -350,6 +353,9 @@ def add_route(self, uri_template, resource, *args, **kwargs):
                                **kwargs)
 
+    def add_static_route(self, prefix, folder):
+        self._static_routes.add(prefix, Statics(prefix, folder))
```
I think it could be helpful to always insert each new prefix at the head of the list. This would ensure a predictable behavior, which we can (and should) document and include in our tests. With this behavior in place, we can note in the docs that duplicate prefixes will override previous ones, and that a more specific prefix may be added _after_ a more general one (e.g., when "/foo/bar" is added after "/foo", a resource such as "/foo/bar/thing.js" would be mapped to the "/foo/bar" route, while "/foo/tmp/thingtoo.js" would be mapped to the "/foo" route.

```py
self._static_routes.insert(0, Statics(prefix, folder))

kgriffs replied:
We should also take the opportunity to verify that the prefix begins with a forward slash, and normalize the string so that it always ends with one as well (so that we match on the complete path segment; matching on the complete segment should also be verified in the unit or functional tests).

kgriffs commented on tests/test_statics.py:
```diff
@@ -0,0 +1,29 @@
+
+    req = Request(testing.create_environ(
+        host='test.com',
+        path='/static/../css//test.css',
```
We'll want to try lots of crazy paths to ensure a malicious client can't break out. You can use ```pytest.mark.parametrize``` to make tests like this a little easier.

kgriffs commented on tests/test_statics.py:
```diff
@@ -0,0 +1,29 @@
+from falcon.statics import Statics
+import falcon.testing as testing
+
+
```
This test module looks like a good start. Once you have the unit tests fleshed out, it would be great if you could also add a few functional tests ala `test_sinks.py` just to make sure everything is working end-to-end. Along the way, it may make sense to allow overriding the statics type (similar to ```API._request_type```). This would make testing easier while also providing a way for developers to customize the behavior.

kgriffs commented on falcon/statics.py:
```diff
@@ -0,0 +1,31 @@
+        stripped = req.path[len(self.prefix):]
+        normalized = os.path.normpath(stripped).lstrip('/')
+
+        if normalized.startswith('../'):
```
Also ensure it does not start with `'/'`. Assuming `self.prefix` ends in a forward slash, there should be no reason why the normalized path should start with one. And if it does (and there should be a test case for this), you end up jumping out of `self.parent_folder`:

```bash
In [8]: os.path.join('/usr/share/nginx/www', '/etc')
Out[8]: '/etc'

```

kgriffs commented:
I went ahead and finished the work on this so we could include it in 1.4. Please review #1157
